ansible_base/lib/jwt_auth/common/auth.py

```diff
@@ -0,0 +1,197 @@
+import logging
+from urllib.parse import urljoin, urlparse
+
+import jwt
+import requests
+from django.conf import settings
+from django.contrib.auth import get_user_model
+from rest_framework.authentication import BaseAuthentication
+from rest_framework.exceptions import AuthenticationFailed
+
+logger = logging.getLogger("ansible_base.lib.jwt_auth.common.auth")
+
+
+default_mapped_user_fields = [
+    "first_name",
+    "last_name",
+    "email",
+    "is_superuser",
+    "is_system_auditor",
+]
+
+
+class JWTCommonAuth:
+    def __init__(self, user_fields=default_mapped_user_fields) -> None:
+        self.mapped_user_fields = user_fields
+
+    def parse_jwt_token(self, request):
+        logger.debug("Starting JWT Authentication")
+        token = request.headers.get("X-DAB-JW-TOKEN", None)
+        if not token:
+            logger.info("X-DAB-JW-TOKEN header not set for JWT authentication")
+            return None, None
+        logger.debug(f"Received token: {token}")
+
+        jwt_key_setting = None
+        try:
+            jwt_key_setting = settings.ANSIBLE_BASE_JWT_KEY
+        except NameError:
+            logger.info("Failed to get the setting ANSIBLE_BASE_JWT_KEY")
+            return None, None
+
+        decryption_key = self.get_decryption_key(
+            jwt_key_setting,
+            validate_certs=getattr(settings, "ANSIBLE_BASE_JWT_VALIDATE_CERT", True),
+            timeout=getattr(settings, "ANSIBLE_BASE_JWT_URL_TIMEOUT", 30),
+        )
+        validated_body = self.validate_token(token, decryption_key)
+        user_model = get_user_model()
+        user, created = user_model.objects.update_or_create(username=validated_body["sub"])
```

Comment:
I'm a little confused here by the lack of `defaults` and `create_defaults`. Sure, this could work if you used those, but the way it looks like it could make a new user with the desired username, but blank first_name, email, etc. and that can't possibly be intended.

Reply:
This probably feels weird at first. 
In here we just map the sub because thats the actual username. But if you look above in this objects init you will see it takes a parameter `user_fields=default_mapped_user_fields` and default_mapped_user_fields is:
```
    "first_name", 
    "last_name", 
    "email", 
    "is_superuser", 
    "is_system_auditor", 
``` 

So this function only extracts the `sub` field and maps that into the `username` creating a user with just a username. However, this object `JWTCommonAuth` is called from `JWTAuthentication` below. Here we initialize with our `JWTCommonAuth` with `map_fields` (which is `default_mapped_fields`) and then, after we auth and get a user with just a username we call `map_user_fields` which will populate the remaining fields.

I did this because we don't have a common user model. That being said, its likely safe to assume that we will all at a minimum inherit form django.contri.auth.models.User which means we could add: first_name, last_name, email, is_superuser to the user create and truncate the `default_mapped_fields` to just `is_system_auditor`. 

LMK if you agree with that assumption or if we should leave it as is.

Reply:
> its likely safe to assume that we will all at a minimum inherit form django.contri.auth.models.User

I strongly believe that to be true. I have been somewhat trying to police the model definitions. eda-server adds `modified_at` which is weird but ok. Of course it uses the AbstractUser base. galaxy_ng uses a custom user model but _does no customizations_ which is for the better, since AWX, much to my chagrin uses `auth.User` itself. In fact, the only sever that does potentially meaningful or dangerous customization is Gateway which adds `is_system_auditor` which has a somewhat unclear/concerning support status in other servers.

So yes I would much prefer to set all AbstractUser fields on creation. You could get it from the abstract model's _meta concrete_fields or however makes sense.